Svc/FrameAccumulator/FrameAccumulator.cpp

```diff
@@ -0,0 +1,176 @@
+        const Fw::SerializeStatus status =
+                this->m_inRing.serialize(&bufferData[offset], serSize);
+        // If data does not fit, there is a coding error
+        FW_ASSERT(
+                status == Fw::FW_SERIALIZE_OK,
+                static_cast<FwAssertArgType>(status),
+                static_cast<FwAssertArgType>(offset),
+                static_cast<FwAssertArgType>(serSize));
+        // Process the data
+        this->processRing();
+        // Update buffer offset and remaining
+        offset += serSize;
+        remaining -= serSize;
+    }
+    // Either all the bytes from the data buffer must be processed, or the ring must be full
+    FW_ASSERT(remaining == 0 || this->m_inRing.get_free_size() == 0, static_cast<FwAssertArgType>(remaining));
+}
+
+    void FrameAccumulator ::processRing() {
+        FW_ASSERT(this->m_detector != nullptr);
+
+        // The number of remaining bytes in the ring buffer
+        U32 remaining = 0;
+        // The protocol status
+        FrameDetector::Status status = FrameDetector::Status::FRAME_DETECTED;
+        // The ring buffer capacity
+        const NATIVE_UINT_TYPE ringCapacity = this->m_inRing.get_capacity();
+
+        // Process the ring buffer looking for at least the header
+        for (U32 i = 0; i < ringCapacity; i++) {
+            // Get the number of bytes remaining in the ring buffer
+            remaining = this->m_inRing.get_allocated_size();
+            // If there are none, we are done
+            if (remaining == 0) {
+                break;
+            }
+            // size_out is a return variable we initialize to zero, but it should be overwritten
+            FwSizeType size_out = 0;
+            // Attempt to detect the frame without changing the circular buffer
+            status = this->m_detector->detect(this->m_inRing, size_out);
+            // Detect must not consume data in the ring buffer
+            FW_ASSERT(
+                    m_inRing.get_allocated_size() == remaining,
+                    static_cast<FwAssertArgType>(m_inRing.get_allocated_size()),
+                    static_cast<FwAssertArgType>(remaining)
+            );
+            // On successful detection, consume data from the ring buffer and place it into an allocated frame
+            if (status == FrameDetector::FRAME_DETECTED) {
+                // size_out must be set to the size of the buffer and must fit within the existing data
+                FW_ASSERT(size_out != 0);
+                FW_ASSERT(
+                        size_out <= remaining,
+                        static_cast<FwAssertArgType>(size_out),
+                        static_cast<FwAssertArgType>(remaining));
+                Fw::Buffer buffer = this->frameAllocate_out(0, size_out);
+                if (buffer.isValid()) {
+                    // Copy out data and rotate
+                    FW_ASSERT(this->m_inRing.peek(buffer.getData(), size_out) == Fw::SerializeStatus::FW_SERIALIZE_OK);
```

Comment:
I'm worried about the performance overhead of this additional copy. In the case that a custom protocol requires _transformation_ of the received data, the decision to split out FrameAccumulator and Deframer means that we'll have to yet another copy in what is already a many-copy architecture.

Here are the copies:

1. The driver loads the received data bytes from the device into a buffer and passes the buffer to FrameAccumulator.
2. FrameAccumulator copies the bytes into its internal ring buffer.
3. FrameAccumulator copies the bytes out of its internal ring buffer into a new allocated frame. (This is the extra copy added by the new model, compared to the existing Deframer.) This frame is passed to the custom deframer.
4. The custom deframer copies the bytes into a new buffer (or reuses the existing buffer) as it translates the received data. The resulting decoded packet is passed onward.
5. The receiving component processes the data and copies it to its final destination, for example saving the data into the filesystem.

I like the idea of splitting out the Router from the Deframer, but I'm not so sure I like the idea of splitting out the FrameAccumulator from the Deframer, at least not in this way.

Also, while I know that F Prime does not discard the condition of an assertion when assertions are disabled, it still unnerves me that we have an assertion condition with a side effect, because that's dangerous in many frameworks.